Bidiagonalization-Symbolic Matrix

sudoWin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
#18797 

#### Brief description of what is fixed or changed
Added symbolic matrix support, rearranged tests

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->
<!-- BEGIN RELEASE NOTES -->
* matrices
  * Added Symbolic Support in House Holder Bidiagonalization
<!-- END RELEASE NOTES -->

sudoWin committed:
Reordering of Tests to increase readibility

sudoWin committed:
NaN Comparision Bug Fix

sudoWin committed:
Removed comparision, added support for symbolic matrices

oscarbenjamin commented on sympy/matrices/tests/test_eigen.py:
```diff
@@ -305,52 +299,47 @@ def test_bidiagonalize():
             comp = random.randint(-1000000000, 1000000000)
             value = real + comp * I
             test_values = test_values + [value]
```
There is already a randMatrix function that you can use

oscarbenjamin commented on sympy/matrices/tests/test_eigen.py:
```diff
@@ -305,52 +299,47 @@ def test_bidiagonalize():
+        M2.simplify()
+        M4.simplify()
+        assert M == M4
+        assert M2 == M0
```
All this code is duplicated above and below. You should just make a `test_matrix` function and use that.

oscarbenjamin commented:
There should be some non-random tests for basic cases.

sudoWin commented:
Can you please suggest which basic cases to add in particular, or I can add any?

oscarbenjamin commented on sympy/matrices/eigen.py:
```diff
@@ -375,27 +375,21 @@ def _is_diagonalizable(M, reals_only=False, **kwargs):
-    v_plus[0, 0] = x[0, 0] + q * norm_x
-    v_minus[0, 0] = x[0, 0] - q * norm_x
+    if x[0, 0] != 0:
+        q = x[0, 0] / abs(x[0, 0])
```
In general you can't assume that `a != 0` means that `a` is nonzero e.g.:
```julia
In [49]: e = cos(1)**2 + sin(1)**2 - 1                                                                                                         

In [50]: e == 0                                                                                                                                
Out[50]: False

In [51]: simplify(e)                                                                                                                           
Out[51]: 0
```
Zero-testing is hard and is something that we should avoid needing to do if at all possible.

sudoWin replied:
Couldn't resolve, Justification is given below

oscarbenjamin commented:
> Can you please suggest which basic cases to add in particular, or I can add any?

At the moment there are only 2 non-random tests. One is for the identity matrix and the other is for a rectangular matrix of floats. The identity matrix is a reasonable test but the output is trivial and the other test does not actually test what the output should be. Neither of those tests actually asserts anything about the matrices produced by this method except that multiplying them gives back the original matrix (after simplification).

I suggest to add tests for some small exact cases like `Matrix([[1, 2], [3, 4]])`. The test should assert exactly what each of the component matrices should be.

There should also be a test for a rectangular case and something involving symbols would also be good.

sudoWin commented:
Thanks for clarifying, I will add and submit it immediately.

sudoWin committed:
Replaced duplicated code with a function, used randMatrix instead of importing random and generating one

sudoWin committed:
Added Symbolic Matrix Test

sudoWin committed:
More tests added

sudoWin commented:
I ran into a small problem, just like other matrix decompositions (QR, LDL), bidiagonalization is not unique, It just guarantees the following,

A = U *  B * V.H
where U and V are orthogonal,
and B is a matrix with non-zero entries at diagonal, and on diagonal above or below.

I also checked the tests for QR decomposition and found that no exact test has been added, and I believe that the reason behind that is the same.
Since the decomposition isn't unique, how can we add exact tests?

sudoWin committed:
Removed an accidentaly added tab character

oscarbenjamin commented:
> Since the decomposition isn't unique, how can we add exact tests?

Presumably the implementation that you have here gives a unique result.

sudoWin commented:
Okay, so I will add tests according to my implementation of bidiagonalization.

sylee957 commented:
I think that a series of tests should be added to check `U, V` are actually orthogonal and `B` is actually bidiagonal, if you hadn't added them, you should add more validity checks.

sudoWin committed:
Added exact test

sudoWin committed:
Tests Changed

sudoWin committed:
Added Exact Tests

sudoWin committed:
Reduced cluttering in test, fixed orthogonal check

sudoWin committed:
Added test_lower_bidiagonal, and test_upper_bidiagonal, to ensure that output is correct

sudoWin commented:
Except for removing the zero testing, the code now has the following:

- [x] Orthogonality Check in Tests
- [x] Upper and Lower Bidiagonal Check in Tests
- [x] Rearranged Tests to reduce cluttering
- [x] Added Symbolic Support
- [x]  Exact Tests added
- [x] Removed a NaN comparison bug
- [x] Used randMatrix instead of importing random

- [ ] Remove zero testing

Well, I am trying if I can remove that without breaking the algorithm

sudoWin commented:
I implemented the direct formula, but I was getting exactly the same results without any difference.

sudoWin commented:
While thinking about removing zero testing, what I believe would be the easiest way is to break the problem into subproblems and address the smallest non-trivial subproblem.

A householder matrix is basically a rotation/reflection and the problem of removing zero testing in its simplest case breaks down to a very simple problem.

If we have a point (x,y) in 2D where x and y are real, can we find the angle by which we need to rotate the line joining origin and this point, so that the line lies on x-Axis, without checking x and y to be zero and also without encountering an expression involving division by zero?

Well. we can't.
The simplest answer may seem to be atan(y/x) but it would fail, if x == 0, and according to the problem, we can't zero test x, nor encounter a division by zero.
The usage of limits should also end up in the same way, as it will still fail on the case when x == 0

So in my opinion, we can't remove the zero testing. Please correct me if the above subproblem has a solution without zero testing and division by zero, or, there is any other approach that can find out the desired answer.

sudoWin commented:
Please review.
Ping @sylee957 @oscarbenjamin.
Thanks!

oscarbenjamin commented:
If zero testing is needed then that's fine but it's tricky and needs to be done carefully. We can not presume that `a != 0` means that a is not zero. It only means that as an expression it is not the literal integer zero. If we use `a.is_zero` then that tries to answer whether the expression is mathematically zero and gives True, False or None (when indeterminate) so we should think about how to handle each of the three cases. One possibility in the None case is to try simplification. Actually I wonder if some kind of simplification might be beneficial as part of the algorithm in general.

Why does this example hang?
```
In [1]: M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])                                                                                                         

In [2]: M.bidiagonal_decomposition()                                                                                                                          
^C
```
This is in the stacktrace:
```
~/current/sympy/sympy/sympy/matrices/eigen.py in _householder_vector(x)
    379     v = x.copy()
    380     if x[0, 0] != 0:
--> 381         q = x[0, 0] / abs(x[0, 0])
    382     else:
    383         q = 1
```
I think that `abs` is trying to determine whether `x[0, 0]` is zero and finding it hard possibly because it is a complicated expression that is actually equal to zero.

From the debugger I see that this is a large expression:
```
ipdb> p x[0, 0].count_ops()                                                                                                                                   
7776
```
I wonder if there is some way to use simplification within the algorithm to avoid creating such complex expressions.

sudoWin committed:
Added QR optimization,Optimized The Functions, changed to is_zero

sudoWin committed:
Added as_mutable() at appropriate places

sudoWin committed:
Modified Tests Accordingly to updated algorithm

sudoWin committed:
Speedup bidiagonalize, Added LHC in bidiagonalize, Added 2 tests for LHC

sudoWin commented:
The following things have been done:

- [x] Changes Switched from ``!=0`` to ``.is_zero``
- [x] Added Simplification as a part of the algorithm when returning householder vector
- [x] Tried Speeding up the code also added a new method
- [x] Since abs was trying to determine whether the expression was zero, removed abs and added equivalent code which doesn't do this
- [x] Added LHC-Bidiagonalization to bidiagonalize

The only thing that remains, is to add LHC bidiagonalization to bidiagonal_decomposition,
I tried adding that but found that it requires a different version of QR decomposition which is not present in Sympy as of now and would involve implementing a whole new algorithm

sudoWin commented:
Please review @oscarbenjamin  @sylee957

oscarbenjamin commented:
This case still hangs with the PR. By adding a few cancels into the algorithm I can get it to compute:
```julia
In [12]: M = Matrix([[1, 2, 3], [3, 4, 5], [6, 7, 8]])

In [13]: a, b, c = M.bidiagonal_decomposition()                                                                                                               

In [14]: a                                                                                                                                                    
Out[14]: 
⎡ -√46    -(-1653570 - 62388⋅√437 + 123120⋅√46 + 191484⋅√38)     -104215284⋅√437 - 2010549600 + 371461362⋅√38 + 362401725⋅√46  ⎤
⎢ ─────   ───────────────────────────────────────────────────    ────────────────────────────────────────────────────────────  ⎥
⎢   46      -91865⋅√437 - 1514642 + 157320⋅√38 + 202122⋅√46      -241601150⋅√437 - 4705177252 + 670183200⋅√38 + 660030132⋅√46  ⎥
⎢                                                                                                                              ⎥
⎢-3⋅√46    -(-734920 - 27728⋅√437 + 54720⋅√46 + 85104⋅√38)     -(-173692140⋅√437 - 3350916000 + 619102270⋅√38 + 604002875⋅√46) ⎥
⎢───────   ────────────────────────────────────────────────    ────────────────────────────────────────────────────────────────⎥
⎢   46     -91865⋅√437 - 1514642 + 157320⋅√38 + 202122⋅√46       -241601150⋅√437 - 4705177252 + 670183200⋅√38 + 660030132⋅√46  ⎥
⎢                                                                                                                              ⎥
⎢-3⋅√46       -643055 - 24262⋅√437 + 47880⋅√46 + 74466⋅√38        -34738428⋅√437 - 670183200 + 123820454⋅√38 + 120800575⋅√46   ⎥
⎢───────    ───────────────────────────────────────────────      ────────────────────────────────────────────────────────────  ⎥
⎣   23      -91865⋅√437 - 1514642 + 157320⋅√38 + 202122⋅√46      -120800575⋅√437 - 2352588626 + 335091600⋅√38 + 330015066⋅√46  ⎦

In [15]: b                                                                                                                                                    
Out[15]: 
⎡                                    √86158                                                                                               ⎤
⎢-√46                                ──────                                                               0                               ⎥
⎢                                      23                                                                                                 ⎥
⎢                                                                                                                                         ⎥
⎢      -(-999972⋅√71174 - 642960⋅√86158 + 325804⋅√818501 + 8635310⋅√1873)   -244674⋅√71174 - 157320⋅√86158 + 79718⋅√818501 + 2112895⋅√1873⎥
⎢ 0    ───────────────────────────────────────────────────────────────────  ──────────────────────────────────────────────────────────────⎥
⎢            -9055955⋅√437 - 149311814 + 15508440⋅√38 + 19924974⋅√46           -9055955⋅√437 - 149311814 + 15508440⋅√38 + 19924974⋅√46    ⎥
⎢                                                                                                                                         ⎥
⎣ 0                                     0                                                                 0                               ⎦

In [16]: c                                                                                                                                                    
Out[16]: 
⎡1      0           0     ⎤
⎢                         ⎥
⎢   -28⋅√1873   -33⋅√1873 ⎥
⎢0  ──────────  ──────────⎥
⎢      1873        1873   ⎥
⎢                         ⎥
⎢   -33⋅√1873    28⋅√1873 ⎥
⎢0  ──────────   ──────── ⎥
⎣      1873        1873   ⎦

In [18]: cancel(a*b*c)                                                                                                                                            
Out[18]: 
⎡1  2  3⎤
⎢       ⎥
⎢3  4  5⎥
⎢       ⎥
⎣6  7  8⎦
```

sylee957 commented:
If `cancel` work with algebraic elements, I'd use cancel selectively for matrix with algebraic elements.
But I still don't get whether `cancel` is actually the right method to use for the purpose like this. Cancel should have a purpose for polynomial factorization.

sudoWin commented:
Can we use ``dotprodsimp`` for that?

sudoWin committed:
Added cancel simplification in algo

sudoWin committed:
Fixed Import Issue

sudoWin commented:
I tried various simplification methods and used ratsimp, as it provided the least execution times for multiple cases

sudoWin committed:
Added tests for LHC bidiagonalize()

oscarbenjamin commented:
This is looking better. I think maybe ratsimp is worth using on the U and V matrices as well (or perhaps just U). With that we get:
```julia
In [13]: M = Matrix([[1, 2], [3, 4]])                                                                                                                         

In [14]: M.bidiagonal_decomposition()                                                                                                                         
Out[14]: 
⎛⎡ -√10    -3⋅√10 ⎤  ⎡      -7⋅√10 ⎤        ⎞
⎜⎢ ─────   ───────⎥  ⎢-√10  ───────⎥        ⎟
⎜⎢   10       10  ⎥  ⎢         5   ⎥  ⎡1  0⎤⎟
⎜⎢                ⎥, ⎢             ⎥, ⎢    ⎥⎟
⎜⎢-3⋅√10     √10  ⎥  ⎢       -√10  ⎥  ⎣0  1⎦⎟
⎜⎢───────    ───  ⎥  ⎢ 0     ───── ⎥        ⎟
⎝⎣   10       10  ⎦  ⎣         5   ⎦        ⎠

In [15]: prod(_)                                                                                                                                              
Out[15]: 
⎡1  2⎤
⎢    ⎥
⎣3  4⎦
```
Whereas the PR currently gives
```julia
In [1]: M = Matrix([[1, 2], [3, 4]])                                                                                                                          

In [2]: M.bidiagonal_decomposition()                                                                                                                          
Out[2]: 
⎛⎡        -√10                ⎛  1   √10⎞ ⎛√10    ⎞   ⎤                         ⎞
⎜⎢        ─────              -⎜- ─ + ───⎟⋅⎜─── + 1⎟   ⎥  ⎡      -7⋅√10 ⎤        ⎟
⎜⎢          10                ⎝  3    3 ⎠ ⎝ 10    ⎠   ⎥  ⎢-√10  ───────⎥        ⎟
⎜⎢                                                    ⎥  ⎢         5   ⎥  ⎡1  0⎤⎟
⎜⎢                                     2              ⎥, ⎢             ⎥, ⎢    ⎥⎟
⎜⎢ ⎛  1   √10⎞ ⎛√10    ⎞    ⎛  1   √10⎞  ⎛√10    ⎞    ⎥  ⎢       -√10  ⎥  ⎣0  1⎦⎟
⎜⎢-⎜- ─ + ───⎟⋅⎜─── + 1⎟  - ⎜- ─ + ───⎟ ⋅⎜─── + 1⎟ + 1⎥  ⎢ 0     ───── ⎥        ⎟
⎝⎣ ⎝  3    3 ⎠ ⎝ 10    ⎠    ⎝  3    3 ⎠  ⎝ 10    ⎠    ⎦  ⎣         5   ⎦        ⎠

In [3]: prod(_)                                                                                                                                               
Out[3]: 
⎡                                                 ⎛  1   √10⎞ ⎛√10    ⎞                       ⎤
⎢                                             √10⋅⎜- ─ + ───⎟⋅⎜─── + 1⎟                       ⎥
⎢                                                 ⎝  3    3 ⎠ ⎝ 10    ⎠   7                   ⎥
⎢            1                                ───────────────────────── + ─                   ⎥
⎢                                                         5               5                   ⎥
⎢                                                                                             ⎥
⎢                                 ⎛             2              ⎞                              ⎥
⎢                                 ⎜  ⎛  1   √10⎞  ⎛√10    ⎞    ⎟         ⎛  1   √10⎞ ⎛√10    ⎞⎥
⎢                             √10⋅⎜- ⎜- ─ + ───⎟ ⋅⎜─── + 1⎟ + 1⎟   7⋅√10⋅⎜- ─ + ───⎟⋅⎜─── + 1⎟⎥
⎢    ⎛  1   √10⎞ ⎛√10    ⎞        ⎝  ⎝  3    3 ⎠  ⎝ 10    ⎠    ⎠         ⎝  3    3 ⎠ ⎝ 10    ⎠⎥
⎢√10⋅⎜- ─ + ───⎟⋅⎜─── + 1⎟  - ────────────────────────────────── + ───────────────────────────⎥
⎣    ⎝  3    3 ⎠ ⎝ 10    ⎠                    5                                 5             ⎦
```

sudoWin committed:
U and V simplification

sudoWin committed:
Indentation Fix

sudoWin commented:
Added ``ratsimp``, but this approximately doubles the time cost of computation of bidiagonal_decomposiiton, should we also a parameter which gives the user the choice of whether to return simplified output or not?

oscarbenjamin commented:
In which case is it slower?

sudoWin commented:
For 3x3 Matrix
``Matrix(3, 3, range(1, 10))``
without simplification - 2.5 seconds
ratsimp - 5.07 seconds
simplify - 8.7 seconds

oscarbenjamin commented:
> For 3x3 Matrix
> `Matrix(3, 3, range(1, 10))`
> without simplification - 2.5 seconds
> ratsimp - 5.07 seconds
> simplify - 8.7 seconds

I'm not sure what you mean. When I commented out the calls to ratsimp it became much slower (it possibly hangs).

sudoWin commented:
```python
def _bidiagonal_decmp_hholder(M, only_b=False):
    m = M.rows
    n = M.cols
    A = M.as_mutable()
    if not only_b:
        U, V = A.eye(m), A.eye(n)
    for i in range(min(m, n)):
        v, bet = _householder_vector(A[i:, i])
        hh_mat = A.eye(m - i) - bet * v * v.H
        hh_mat = _simplify(hh_mat)
        A[i:, i:] = hh_mat * A[i:, i:]
        A[i:, i:] = A[i:, i:].applyfunc(ratsimp)
        if not only_b:
            temp = A.eye(m)
            temp[i:, i:] = hh_mat
            U = U * temp
            U = U.applyfunc(ratsimp)
        if i + 1 <= n - 2:
            v, bet = _householder_vector(A[i, i+1:].T)
            hh_mat = A.eye(n - i - 1) - bet * v * v.H
            hh_mat = _simplify(hh_mat)
            A[i:, i+1:] = A[i:, i+1:] * hh_mat
            A[i:, i+1:] = A[i:, i+1:].applyfunc(ratsimp)
            if not only_b:
                temp = A.eye(n)
                temp[i+1:, i+1:] = hh_mat
                V = temp * V
                V = V.applyfunc(ratsimp)
    if not only_b:
        return U, A, V
    else:
        return A
```
The above code takes 5.05 seconds

When we remove simplification for U and V
```python3
def _bidiagonal_decmp_hholder(M, only_b=False):
    m = M.rows
    n = M.cols
    A = M.as_mutable()
    if not only_b:
        U, V = A.eye(m), A.eye(n)
    for i in range(min(m, n)):
        v, bet = _householder_vector(A[i:, i])
        hh_mat = A.eye(m - i) - bet * v * v.H
        hh_mat = _simplify(hh_mat)
        A[i:, i:] = hh_mat * A[i:, i:]
        A[i:, i:] = A[i:, i:].applyfunc(ratsimp)
        if not only_b:
            temp = A.eye(m)
            temp[i:, i:] = hh_mat
            U = U * temp
        #    U = U.applyfunc(ratsimp)
        if i + 1 <= n - 2:
            v, bet = _householder_vector(A[i, i+1:].T)
            hh_mat = A.eye(n - i - 1) - bet * v * v.H
            hh_mat = _simplify(hh_mat)
            A[i:, i+1:] = A[i:, i+1:] * hh_mat
            A[i:, i+1:] = A[i:, i+1:].applyfunc(ratsimp)
            if not only_b:
                temp = A.eye(n)
                temp[i+1:, i+1:] = hh_mat
                V = temp * V
        #        V = V.applyfunc(ratsimp)
    if not only_b:
        return U, A, V
    else:
        return A
```
Then it takes only 2.5 seconds
so I was wondering whether we should add a parameter to bypass simplification

oscarbenjamin commented:
These are the timings I get when calling radsimp for U and V as in this PR currently:
```
In [1]: M = Matrix(3, 3, range(1, 10))                                                                                                                        

In [2]: %time bd = M.bidiagonal_decomposition()                                                                                                               
CPU times: user 8.87 s, sys: 30.5 ms, total: 8.9 s
Wall time: 8.92 s

In [3]: %time cancel(prod(bd))                                                                                                                                
CPU times: user 7.34 s, sys: 21.4 ms, total: 7.36 s
Wall time: 7.39 s
Out[3]: 
⎡1  2  3⎤
⎢       ⎥
⎢4  5  6⎥
⎢       ⎥
⎣7  8  9⎦
```
These are the timings I get if I comment out those two radsimp calls:
```
In [1]: M = Matrix(3, 3, range(1, 10))                                                                                                                        

In [2]: %time bd = M.bidiagonal_decomposition()                                                                                                               
CPU times: user 5.94 s, sys: 22.9 ms, total: 5.96 s
Wall time: 5.97 s

In [3]: %time cancel(prod(bd))                                                                                                                                
CPU times: user 15.3 s, sys: 34.4 ms, total: 15.3 s
Wall time: 15.4 s
Out[3]: 
⎡1  2  3⎤
⎢       ⎥
⎢4  5  6⎥
⎢       ⎥
⎣7  8  9⎦
```
What I see is that not simplifying internally makes the result return quicker but then leads to subsequent calculations slowing down because of the large complicated expressions that result.

There is a tradeoff here. It's hard to answer precisely what we should do unless we have a clear idea what the bidiagonal decomposition could be used for.

sylee957 commented:
In my opinion, if the U, V are simply constructed using the matrix multiplication, we can look for some alternative lazy evaluated form like the idea in #18867.
For example, lots of householder-based methods store the information about the householder vector in alternative ways. And usually getting a chain of reflection vectors used in the transformation is more important than just getting the decomposition.
With lazy evaluation capability of matrix expressions, it can make the needs of the both worlds meet.

sudoWin commented:
That looks great if we lazy evaluate the expression, 

I think we should replace U by a lazy evaluated Un,.....U3, U2, U1 and V by V1, V2, V3,...., Vn?

sylee957 commented:
You can wrap like MatMul(U3, U2, U1) such that it looks like `U` when multiplication is done.

sudoWin commented:
I have a doubt, if we have a Lazy evaluated expression, say X =  A + B*C, how to access a particular matrix, say B in it?

sylee957 commented:
You can use `.args` and nest through the tree structure.

sudoWin commented:
Thanks, I will submit a commit soon.

oscarbenjamin commented:
> You can use `.args` and nest through the tree structure.

This should be made clear in the docs

sudoWin committed:
Switched to Lazy Evaluation of Matrix Expressions

sudoWin committed:
Modified Tests for Lazy Evaluation of U, and V in bidiagonal_decomposition

sudoWin commented:
Please review @oscarbenjamin @sylee957

sylee957 commented:
Why not construct the U, V in flattened forms in the first place? I see MatMul is nested.

sudoWin committed:
Changed from nested MatMul to Flattened MatMul

sudoWin commented:
Shifted from nested to flattened MatMul

sylee957 commented on sympy/matrices/eigen.py:
```diff
@@ -456,18 +480,23 @@ def _bidiagonal_decomposition(M, upper=True):
-    return _bidiagonal_decmp_hholder(M)
+        X = M.H.bidiagonal_decomposition()
+        return X[2].T.C, X[1].T.C, X[0].T.C
+    if method == "lhc":
```
When `_bidiagonal_decmp_hholder` is computed after the QR decomposition, then the original `Q` from the QR decomposition can be multiplied to the first factor to give a new orthogonal matrix for the decomposition.
I'd expect that when bidiagonal decomposition is done after the QR decomposition, the first matrix should be identity matrix, so it will be easier.

sudoWin replied:
> When _bidiagonal_decmp_hholder is computed after the QR decomposition, then the original Q from the QR decomposition can be multiplied to the first factor to give a new orthogonal matrix for the decomposition.

This can be done but there is a problem, The dimensions of the bidiagonalized matrix B are not the same as the dimensions of the input Matrix. Now when we just have to return the bidiagonalized matrix, it is not a big problem, as we can pad zeroes accordingly and the answer remains correct. But when we have to return the bidiagonalization, padding zeroes, is not the right way because even if it keeps everything right, the final Matrix U, which is ``Q * _U`` loses its orthogonality.

> I'd expect that when bidiagonal decomposition is done after the QR decomposition, the first matrix should be the identity matrix, so it will be easier.

Sadly this doesn't hold for cases of 3x3 or bigger,  only U1 is identity and U2,U3,.....Un need not be an identity matrix.

sudoWin replied:
Solution: I guess there are probably two different versions of QR, though I am not totally sure.
Here is something I did in Sympy and the equivalent in Matlab
```python

In [5]: A = Matrix(3, 3, range(1, 10))

In [6]: Q, R = A.QRdecomposition()

In [7]: Q.evalf(5)
Out[7]:
⎡0.12309  0.90453 ⎤
⎢                 ⎥
⎢0.49237  0.30151 ⎥
⎢                 ⎥
⎣0.86164  -0.30151⎦

In [8]: R.evalf(5)
Out[8]:
⎡8.124  9.6011   11.078⎤
⎢                      ⎥
⎣  0    0.90453  1.8091⎦

In [9]: Q * Q.T
Out[9]:
⎡5/6   1/3  -1/6⎤
⎢               ⎥
⎢1/3   1/3  1/3 ⎥
⎢               ⎥
⎣-1/6  1/3  5/6 ⎦

In [10]: Q.T * Q
Out[10]:
⎡1  0⎤
⎢    ⎥
⎣0  1⎦

```

Matlab:
```python
>> A = [ 1 2 3;4 5 6;7 8 9]

A =

     1     2     3
     4     5     6
     7     8     9

>> [Q,R] = qr(A)

Q =

   -0.1231    0.9045    0.4082
   -0.4924    0.3015   -0.8165
   -0.8616   -0.3015    0.4082


R =

   -8.1240   -9.6011  -11.0782
         0    0.9045    1.8091
         0         0   -0.0000

>> Q * Q.'

ans =

    1.0000    0.0000    0.0000
    0.0000    1.0000    0.0000
    0.0000    0.0000    1.0000

>> Q.' * Q

ans =

    1.0000   -0.0000   -0.0000
   -0.0000    1.0000    0.0000
   -0.0000    0.0000    1.0000
```

For a Matrix M to be orthogonal, 
M.T * M = I
M * M.T = I
and that I believe is a bug somewhere in QR in sympy, where we are getting correct R, but somehow an incomplete Q.

sylee957 replied:
I think that you should augment the `Q` with any basis that is linearly independent to original matrix and run orthogonalization again. (rather than simply padding with zero)

The idea behind I had made Q to be semi-orthogonal was because of skipping this redundant computation. So I may not consider that this is a bad result.
But the docs may be improved to give some rank-deficient examples.

```python3
from sympy import *

A = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
Q, R = A.QRdecomposition()
Q_aug = Matrix.hstack(Q, Matrix([0, 0, 1])).QRdecomposition()[0]
R_aug = Matrix.vstack(R, Matrix([0, 0, 0]).T)

print(Q_aug * Q_aug.T)
print(Q_aug.T * Q_aug)
print(Q_aug * R_aug)
```

sudoWin replied:
That's really great! I will submit a commit, just one problem remains, how can I find a basis linearly independent to the original matrix? Is there any function/or anything for that?

sylee957 replied:
Usually, you can try padding every standard basis
`Matrix([1, 0, 0])`, `Matrix([0, 1, 0])`, `Matrix([0, 0, 1])` and orthogonalize it until succeeds.
And mathematically, it's guaranteed to succeed because there should be any one of three can be linearly independent to the original basis.

But in my opinion, I'd rather have other open questions like whether extending the bidiagonalization to return a semi-orthogonal matrix like is possible or not.

sylee957 commented on sympy/matrices/eigen.py:
```diff
@@ -481,22 +510,36 @@ def _bidiagonalize(M, upper=True):
-    return _eval_bidiag_hholder(M)
-
+        return M.H.bidiagonalize(method=method).H
+    if method == "lhc":
```
Unfortunately, I don't find many things are optimized to `LHC` method you have implemented. 
And if it is basically computing the decomposition after another decomposition, it wouldn't be needed to made into a user API, because it's the matter of the execution order.
I'd expect some optimizations like skipping some eliminations.

sudoWin replied:
Currently, LHC is approximately 13x Faster than householder for a 3x3 Matrix, and I beleive many many times faster than Householder for 4x4 or above matrices.
```python3
In [10]: A = Matrix(3, 3, range(1, 10))

In [11]: %timeit A.bidiagonalize()
187 ms ± 13.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [12]: %timeit A.bidiagonalize(method="householder")
2.49 s ± 133 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
I could figure out only one optimization and that is skipping the very first elimination and  It won't give a big reduction in time cost.
I will add a fix for that in the next commit.
If you beleive there are others, please do tell me.

czgdp1807 commented:
@sudoWin Please resolve the conflicts in this PR. Thanks for your contributions.

oscarbenjamin committed:
Merge branch 'master' into householder-fix
